demos/coreMQTT/mqtt_demo_keep_alive.c

```diff
@@ -0,0 +1,1172 @@
+ * and no control packets have been sent after some given interval.
+ *
+ * @param pxTimer The auto-reload software timer for handling keep alive.
+ */
+static void prvKeepAliveTimerCallback( TimerHandle_t pxTimer );
+
+/**
+ * @brief The application callback function for getting the incoming publish
+ * and incoming acks reported from the MQTT library.
+ *
+ * @param pxMQTTContext MQTT context pointer.
+ * @param pxPacketInfo Packet Info pointer for the incoming packet.
+ * @param pxDeserializedInfo Deserialized information from the incoming packet.
+ */
+static void prvEventCallback( MQTTContext_t * pxMQTTContext,
+                              MQTTPacketInfo_t * pxPacketInfo,
+                              MQTTDeserializedInfo_t * pxDeserializedInfo );
+
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief Static buffer used to hold MQTT messages being sent and received.
+ */
+static uint8_t ucSharedBuffer[ mqttexampleSHARED_BUFFER_SIZE ];
+
+/**
+ * @brief Static buffer used to hold PINGREQ messages to be sent.
+ */
+static uint8_t ucPingReqBuffer[ MQTT_PACKET_PINGREQ_SIZE ];
+
+/**
+ * @brief Global entry time into the application to use as a reference timestamp
+ * in the #prvGetTimeMs function. #prvGetTimeMs will always return the difference
+ * between the current time and the global entry time. This will reduce the chances
+ * of overflow for the 32 bit unsigned integer used for holding the timestamp.
+ */
+static uint32_t ulGlobalEntryTimeMs;
+
+/**
+ * @brief Packet Identifier generated when Subscribe request was sent to the broker;
+ * it is used to match received Subscribe ACK to the transmitted ACK.
+ */
+static uint16_t usSubscribePacketIdentifier;
+
+/**
+ * @brief Packet Identifier generated when Unsubscribe request was sent to the broker;
+ * it is used to match received Unsubscribe response to the transmitted unsubscribe
+ * request.
+ */
+static uint16_t usUnsubscribePacketIdentifier;
+
+/**
+ * @brief A pair containing a topic filter and its SUBACK status.
+ */
+typedef struct topicFilterContext
+{
+    const char * pcTopicFilter;
+    MQTTSubAckStatus_t xSubAckStatus;
+} topicFilterContext_t;
+
+/**
+ * @brief An array containing the context of a SUBACK; the SUBACK status
+ * of a filter is updated when the event callback processes a SUBACK.
+ */
+static topicFilterContext_t xTopicFilterContext[ mqttexampleTOPIC_COUNT ] =
+{
+    { mqttexampleTOPIC, MQTTSubAckFailure }
+};
+
+/**
+ * @brief Timer to handle the MQTT keep-alive mechanism.
+ */
+static TimerHandle_t xKeepAliveTimer;
+
+/**
+ * @brief Storage space for xKeepAliveTimer.
+ */
+static StaticTimer_t xKeepAliveTimerBuffer;
+
+/**
+ * @brief Set to true when PINGREQ is sent then false when PINGRESP is received.
+ */
+static volatile bool xWaitingForPingResp = false;
+
+/**
+ * @brief Set to true when an error occurs in the prvKeepAliveTimerCallback().
+ */
+static volatile bool xKeepAliveError = false;
```

Comment:
What is the purpose of making these _shared_ memory variables `volatile`?

Reply:
Because they are accessed by multiple threads.